// react-native-ui-kitten/ui/radioGroup/radioGroup.component.js
"use strict";
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const react_native_1 = require("react-native");
const theme_1 = require("../../theme");
/**
 * Renders a group of `Radio` buttons.
 *
 * @extends React.Component
 *
 * @property {ReactElement<RadioProps> | ReactElement<RadioProps>[]} children - Determines radio buttons in group.
 *
 * @property {number} selectedIndex - Determines the index of selected button
 *
 * @property {(index: number) => void} onChange - Fires when selected radio is changed.
 *
 * @property {ViewProps} ...ViewProps - Any props applied to View component.
 *
 * @overview-example RadioGroupSimpleUsage
 */
class RadioGroupComponent extends react_1.default.Component {
    constructor() {
        super(...arguments);
        this.onRadioChange = (index) => {
            if (this.props.onChange) {
                this.props.onChange(index);
            }
        };
        this.getComponentStyle = (source) => {
            const { itemMarginVertical, ...containerParameters } = source;
            return {
                container: containerParameters,
                item: {
                    marginVertical: itemMarginVertical,
                },
            };
        };
        this.renderRadioElements = (source, style) => {
            return react_1.default.Children.map(source, (element, index) => {
                return react_1.default.cloneElement(element, {
                    key: index,
                    style: [style, element.props.style],
                    checked: this.props.selectedIndex === index,
                    onChange: () => this.onRadioChange(index),
                });
            });
        };
    }
    render() {
        const { themedStyle, style, children, ...derivedProps } = this.props;
        const componentStyle = this.getComponentStyle(themedStyle);
        const radioElements = this.renderRadioElements(children, componentStyle.item);
        return (react_1.default.createElement(react_native_1.View, Object.assign({}, derivedProps, { style: [componentStyle.container, styles.container, style] }), radioElements));
    }
}
RadioGroupComponent.styledComponentName = 'RadioGroup';
RadioGroupComponent.defaultProps = {
    selectedIndex: -1,
};
const styles = react_native_1.StyleSheet.create({
    container: {},
});
exports.RadioGroup = theme_1.styled(RadioGroupComponent);
