// react-native-ui-kitten/ui/tab/tabView.component.js
"use strict";
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const react_native_1 = require("react-native");
const tabBar_component_1 = require("./tabBar.component");
const viewPager_component_1 = require("../viewPager/viewPager.component");
class TabViewChildElement {
}
class TabViewChildren {
    constructor() {
        this.tabs = [];
        this.content = [];
    }
}
/**
 * `TabView` is a dynamic tabset component. Allows flipping through the tab "pages".
 *
 * @extends React.Component
 **
 * @property {number} selectedIndex - Determines current tab index.
 *
 * @property {StyleProp<ViewStyle>} tabBarStyle - Determines style TabBar component.
 *
 * @property {StyleProp<ViewStyle>} indicatorStyle - Determines style of selected tab indicator.
 *
 * @property {(index: number) => void} onSelect - Fires on "page" select with corresponding index.
 *
 * @property {ReactElement<TabProps> | ReactElement<TabProps>[]} children - Determines children of the component.
 *
 * @property {(index: number) => boolean} shouldLoadComponent - Determines loading behavior particular page and can be
 * used for lazy loading.
 *
 * @property {(offset: number) => void} onOffsetChange - Fires on scroll event with current scroll offset.
 *
 * @property {ViewProps} ...ViewProps - Any props applied to View component.
 *
 * @overview-example TabViewSimpleUsage
 *
 * @overview-example TabViewWithIcon
 *
 * @overview-example TabViewLazyLoading
 *
 * @example TabViewInlineStyling
 */
class TabView extends react_1.default.Component {
    constructor() {
        super(...arguments);
        this.viewPagerRef = react_1.default.createRef();
        this.tabBarRef = react_1.default.createRef();
        this.onBarSelect = (index) => {
            const { current: viewPager } = this.viewPagerRef;
            viewPager.scrollToIndex({ index, animated: true });
        };
        this.onPagerOffsetChange = (offset) => {
            const { current: tabBar } = this.tabBarRef;
            const tabCount = react_1.default.Children.count(tabBar.props.children);
            tabBar.scrollToOffset({ offset: offset / tabCount });
        };
        this.onPagerSelect = (selectedIndex) => {
            if (this.props.onSelect) {
                this.props.onSelect(selectedIndex);
            }
        };
        this.renderComponentChild = (element, index) => {
            return {
                tab: react_1.default.cloneElement(element, { key: index }),
                content: element.props.children,
            };
        };
        this.renderComponentChildren = (source) => {
            const children = react_1.default.Children.toArray(source);
            return children.reduce((acc, element, index) => {
                const { tab, content } = this.renderComponentChild(element, index);
                return {
                    tabs: [...acc.tabs, tab],
                    content: [...acc.content, content],
                };
            }, new TabViewChildren());
        };
    }
    render() {
        const { style, selectedIndex, children, tabBarStyle, indicatorStyle, ...derivedProps } = this.props;
        const { tabs, content } = this.renderComponentChildren(children);
        return (react_1.default.createElement(react_native_1.View, Object.assign({ style: [styles.container, style] }, derivedProps),
            react_1.default.createElement(tabBar_component_1.TabBar, { style: tabBarStyle, ref: this.tabBarRef, selectedIndex: selectedIndex, indicatorStyle: indicatorStyle, onSelect: this.onBarSelect }, tabs),
            react_1.default.createElement(viewPager_component_1.ViewPager, Object.assign({ ref: this.viewPagerRef }, derivedProps, { style: [styles.container, style], selectedIndex: selectedIndex, shouldLoadComponent: this.props.shouldLoadComponent, onOffsetChange: this.onPagerOffsetChange, onSelect: this.onPagerSelect }), content)));
    }
}
exports.TabView = TabView;
TabView.defaultProps = {
    selectedIndex: 0,
};
const styles = react_native_1.StyleSheet.create({
    container: {
        overflow: 'hidden',
    },
});
