// react-native-ui-kitten/theme/theme/withStyles.js
"use strict";
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const hoist_non_react_statics_1 = __importDefault(require("hoist-non-react-statics"));
const themeContext_1 = require("./themeContext");
/**
 * `withStyles` is a High Order Function which is used to create themed style for non-styled component.
 * Basically used when need to use theme variable somewhere.
 * Returns component class which can be used as themed component.
 *
 * @property {ThemeType} theme - Determines theme used to style component.
 *
 * @property {Styles} themedStyle - Determines component style for it's current state.
 *
 * @param Component - Type: {ComponentType}. Determines class of component to be themed.
 *
 * @param createStyles - Type: {(theme: ThemeType) => Styles}. Determines arrow function used to create styles.
 *
 * @overview-example WithStylesSimpleUsage
 */
exports.withStyles = (Component, createStyles) => {
    class Wrapper extends react_1.default.Component {
        constructor() {
            super(...arguments);
            this.withThemedProps = (props, theme) => {
                const themedStyle = createStyles && createStyles(theme);
                return { ...props, theme, themedStyle };
            };
            this.renderWrappedElement = (theme) => {
                const { forwardedRef, ...restProps } = this.props;
                const props = this.withThemedProps(restProps, theme);
                return (react_1.default.createElement(Component, Object.assign({}, props, { ref: forwardedRef })));
            };
        }
        render() {
            return (react_1.default.createElement(themeContext_1.ThemeContext.Consumer, null, this.renderWrappedElement));
        }
    }
    const WrappingElement = (props, ref) => {
        return (
        // @ts-ignore
        react_1.default.createElement(Wrapper, Object.assign({}, props, { forwardedRef: ref })));
    };
    const ThemedComponent = react_1.default.forwardRef(WrappingElement);
    ThemedComponent.displayName = Component.displayName || Component.name;
    hoist_non_react_statics_1.default(ThemedComponent, Component);
    // @ts-ignore
    return ThemedComponent;
};
