// react-native-ui-kitten/ui/select/selectGroupOption.component.js
"use strict";
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const react_native_1 = require("react-native");
const theme_1 = require("../../theme");
const selectOption_component_1 = require("./selectOption.component");
class SelectGroupOptionComponent extends react_1.default.Component {
    constructor() {
        super(...arguments);
        this.getComponentStyle = (source) => {
            const { itemPaddingHorizontal, ...containerParameters } = source;
            return {
                container: containerParameters,
                item: {
                    paddingHorizontal: itemPaddingHorizontal,
                },
            };
        };
        this.getGroupOptionMainItemProps = (children) => {
            const selectedItems = children.map(element => element.props.selected);
            const someSelected = selectedItems.some((item) => item === true);
            const everySelected = selectedItems.every((item) => item === true);
            if (someSelected && !everySelected) {
                return { selected: true, indeterminate: true };
            }
            if (!someSelected) {
                return { selected: false, indeterminate: false };
            }
            if (everySelected) {
                return { selected: true, indeterminate: false };
            }
        };
        this.renderGroupOptionElement = (option, index) => {
            return (react_1.default.createElement(selectOption_component_1.SelectOption, { key: index, multi: this.props.multi, item: option, selected: this.props.isOptionSelected(option), onSelect: this.props.onSelect }));
        };
        this.renderGroupOptionElements = (componentStyle) => {
            return this.props.item.items.map((option, index) => {
                const optionElement = this.renderGroupOptionElement(option, index);
                return react_1.default.cloneElement(optionElement, {
                    style: [optionElement.props.style, componentStyle.item],
                });
            });
        };
        this.renderSingleElement = () => {
            return (react_1.default.createElement(selectOption_component_1.SelectOption, { key: 0, multi: this.props.multi, item: this.props.item, selected: false, onSelect: this.props.onSelect, disabled: true }));
        };
        this.renderGroupElement = (children) => {
            return (react_1.default.createElement(selectOption_component_1.SelectOption, Object.assign({}, this.getGroupOptionMainItemProps(children), { key: 0, multi: this.props.multi, item: this.props.item, onSelect: this.props.onSelect })));
        };
        this.renderComponentChildren = (componentStyle) => {
            const groupOptions = this.renderGroupOptionElements(componentStyle);
            return [
                this.props.multi ? this.renderGroupElement(groupOptions) : this.renderSingleElement(),
                groupOptions,
            ];
        };
    }
    render() {
        const { container, ...componentStyle } = this.getComponentStyle(this.props.themedStyle);
        const [mainElement, groupOptionElements] = this.renderComponentChildren(componentStyle);
        return (react_1.default.createElement(react_native_1.View, { style: container },
            mainElement,
            groupOptionElements));
    }
}
SelectGroupOptionComponent.styledComponentName = 'SelectGroupOption';
exports.SelectGroupOption = theme_1.styled(SelectGroupOptionComponent);
