// react-native-ui-kitten/ui/toggle/toggle.component.js
"use strict";
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const react_native_1 = require("react-native");
const theme_1 = require("../../theme");
const text_component_1 = require("../text/text.component");
const checkmark_component_1 = require("../support/components/checkmark.component");
const services_1 = require("../support/services");
/**
 * Styled `Toggle` component.
 *
 * @extends React.Component
 *
 * @property {boolean} checked - Determines whether component is checked.
 * Default is `false`.
 *
 * @property {boolean} disabled - Determines whether component is disabled.
 * Default is `false.
 *
 * @property {string} status - Determines the status of the component.
 * Can be `basic`, `primary`, `success`, `info`, `warning`, `danger` or `control`.
 * Default is `basic`.
 *
 * @property {string} text - Determines text of the component.
 *
 * @property {StyleProp<TextStyle>} textStyle - Customizes text style.
 *
 * @property {(checked: boolean) => void} onChange - Fires when selection state is changed.
 *
 * @property {TouchableOpacityProps} ...TouchableOpacityProps - Any props applied to TouchableOpacity component.
 *
 * @overview-example ToggleSimpleUsage
 *
 * @overview-example ToggleStates
 *
 * @overview-example ToggleStatus
 *
 * @example ToggleInlineStyling
 */
class ToggleComponent extends react_1.default.Component {
    constructor(props) {
        super(props);
        this.webEventResponder = services_1.WebEventResponder.create(this);
        this.onMouseEnter = () => {
            if (!this.props.disabled) {
                this.props.dispatch([theme_1.Interaction.HOVER]);
            }
        };
        this.onMouseLeave = () => {
            if (!this.props.disabled) {
                this.props.dispatch([]);
            }
        };
        this.onFocus = () => {
            if (!this.props.disabled) {
                this.props.dispatch([theme_1.Interaction.FOCUSED]);
            }
        };
        this.onBlur = () => {
            if (!this.props.disabled) {
                this.props.dispatch([]);
            }
        };
        // PanResponderCallbacks
        this.onStartShouldSetPanResponder = () => {
            return true;
        };
        this.onStartShouldSetPanResponderCapture = () => {
            return true;
        };
        this.onMoveShouldSetPanResponder = () => {
            return true;
        };
        this.onMoveShouldSetPanResponderCapture = () => {
            return true;
        };
        this.onPanResponderTerminationRequest = () => {
            return false;
        };
        this.onPanResponderGrant = () => {
            const { checked, disabled, themedStyle } = this.props;
            if (disabled) {
                return;
            }
            this.onPressIn();
            if (this.thumbTranslateAnimationActive) {
                this.thumbTranslateAnimationActive = false;
                this.stopAnimations();
                return;
            }
            this.animateThumbWidth(themedStyle.thumbWidth * 1.2);
            this.animateEllipseScale(checked ? 1 : 0.01);
        };
        this.onPanResponderMove = () => {
            return true;
        };
        this.onPanResponderRelease = (e, gestureState) => {
            const { checked, disabled, themedStyle } = this.props;
            if (!disabled) {
                if ((!checked && gestureState.dx > -5) || (checked && gestureState.dx < 5)) {
                    this.toggle(this.onPress);
                }
                else {
                    this.animateEllipseScale(checked ? 0.01 : 1);
                }
            }
            this.animateThumbWidth(themedStyle.thumbWidth);
            this.onPressOut();
        };
        this.onPressIn = () => {
            this.props.dispatch([theme_1.Interaction.ACTIVE]);
        };
        this.onPressOut = () => {
            this.props.dispatch([]);
        };
        this.onPress = () => {
            if (this.props.onChange) {
                this.props.onChange(!this.props.checked);
            }
        };
        this.getComponentStyle = (source) => {
            const { checked, disabled } = this.props;
            const { outlineWidth, outlineHeight, outlineBorderRadius, outlineBackgroundColor, thumbWidth, thumbHeight, thumbBorderRadius, thumbBackgroundColor, textMarginHorizontal, textFontSize, textFontWeight, textLineHeight, textFontFamily, textColor, iconWidth, iconHeight, iconTintColor, backgroundColor, borderColor, ...containerParameters } = source;
            return {
                toggleContainer: {},
                ellipseContainer: {
                    borderColor: borderColor,
                    backgroundColor: backgroundColor,
                    ...containerParameters,
                },
                highlight: {
                    width: outlineWidth,
                    height: outlineHeight,
                    borderRadius: outlineBorderRadius,
                    backgroundColor: outlineBackgroundColor,
                },
                ellipse: {
                    width: containerParameters.width - (containerParameters.borderWidth * 2),
                    height: containerParameters.height - (containerParameters.borderWidth * 2),
                    borderRadius: (source.height - (source.borderWidth * 2)) / 2,
                    backgroundColor: backgroundColor,
                },
                thumb: {
                    alignSelf: checked ? 'flex-end' : 'flex-start',
                    width: this.thumbWidthAnimation,
                    height: thumbHeight,
                    borderRadius: thumbBorderRadius,
                    backgroundColor: thumbBackgroundColor,
                    elevation: disabled ? 0 : 5,
                    transform: [{ translateX: this.thumbTranslateAnimation }],
                },
                text: {
                    marginHorizontal: textMarginHorizontal,
                    fontSize: textFontSize,
                    fontWeight: textFontWeight,
                    lineHeight: textLineHeight,
                    fontFamily: textFontFamily,
                    color: textColor,
                },
                icon: {
                    width: source.iconWidth,
                    height: source.iconHeight,
                    fill: iconTintColor,
                    stroke: iconTintColor,
                    strokeWidth: 3,
                },
            };
        };
        this.animateThumbTranslate = (value, callback = () => null) => {
            this.thumbTranslateAnimationActive = true;
            react_native_1.Animated.timing(this.thumbTranslateAnimation, {
                toValue: services_1.I18nLayoutService.select(value, -value),
                duration: 150,
                easing: react_native_1.Easing.linear,
            }).start(() => {
                this.thumbTranslateAnimationActive = false;
                callback();
            });
        };
        this.animateThumbWidth = (value, callback = () => null) => {
            react_native_1.Animated.timing(this.thumbWidthAnimation, {
                toValue: value,
                duration: 150,
                easing: react_native_1.Easing.linear,
            }).start(callback);
        };
        this.animateEllipseScale = (value, callback = () => null) => {
            react_native_1.Animated.timing(this.ellipseScaleAnimation, {
                toValue: value,
                duration: 200,
                easing: react_native_1.Easing.linear,
            }).start(callback);
        };
        this.stopAnimations = () => {
            const value = this.props.checked ? 0.01 : 1;
            this.thumbTranslateAnimation.stopAnimation();
            this.ellipseScaleAnimation.stopAnimation();
            this.thumbWidthAnimation.stopAnimation();
            this.ellipseScaleAnimation.setValue(value);
        };
        this.toggle = (callback = (nextValue) => null) => {
            const value = this.props.checked ? -20 : 20;
            this.animateThumbTranslate(value, () => {
                this.thumbTranslateAnimation.setValue(0);
                callback(!this.props.checked);
            });
            this.animateThumbWidth(this.props.themedStyle.thumbWidth);
        };
        this.renderTextElement = (style) => {
            return (react_1.default.createElement(text_component_1.Text, { style: [style, this.props.textStyle] }, this.props.text));
        };
        this.renderComponentChildren = (style) => {
            return [
                this.props.text && this.renderTextElement(style.text),
            ];
        };
        const { checked, themedStyle } = props;
        this.thumbWidthAnimation = new react_native_1.Animated.Value(themedStyle.thumbWidth);
        this.thumbTranslateAnimation = new react_native_1.Animated.Value(0);
        this.ellipseScaleAnimation = new react_native_1.Animated.Value(checked ? 0.01 : 1.0);
        this.thumbTranslateAnimationActive = false;
        this.panResponder = react_native_1.PanResponder.create(this);
    }
    render() {
        const { themedStyle, style, checked, ...restProps } = this.props;
        const componentStyle = this.getComponentStyle(themedStyle);
        const [textElement] = this.renderComponentChildren(componentStyle);
        return (react_1.default.createElement(react_native_1.View, Object.assign({}, this.panResponder.panHandlers, { style: [styles.container, style] }),
            react_1.default.createElement(react_native_1.TouchableOpacity, Object.assign({ activeOpacity: 1.0 }, restProps, this.webEventResponder.eventHandlers, { style: [componentStyle.toggleContainer, webStyles.toggleContainer, styles.toggleContainer] }),
                react_1.default.createElement(react_native_1.View, { style: [componentStyle.highlight, styles.highlight] }),
                react_1.default.createElement(react_native_1.Animated.View, { style: [componentStyle.ellipseContainer, styles.ellipseContainer] },
                    react_1.default.createElement(react_native_1.Animated.View, { style: [componentStyle.ellipse, styles.ellipse] }),
                    react_1.default.createElement(react_native_1.Animated.View, { style: [componentStyle.thumb, styles.thumb] },
                        react_1.default.createElement(checkmark_component_1.CheckMark, Object.assign({}, componentStyle.icon))))),
            textElement));
    }
}
exports.ToggleComponent = ToggleComponent;
ToggleComponent.styledComponentName = 'Toggle';
const styles = react_native_1.StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
    },
    toggleContainer: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    ellipseContainer: {
        justifyContent: 'center',
        alignSelf: 'center',
        overflow: 'hidden',
    },
    ellipse: {
        alignSelf: 'center',
        position: 'absolute',
    },
    highlight: {
        alignSelf: 'center',
        position: 'absolute',
    },
    thumb: {
        justifyContent: 'center',
        alignItems: 'center',
    },
});
const webStyles = react_native_1.Platform.OS === 'web' && react_native_1.StyleSheet.create({
    toggleContainer: {
        // @ts-ignore
        outlineWidth: 0,
    },
});
exports.Toggle = theme_1.styled(ToggleComponent);
